File: ocaml/bob/bob.ml
open Base

let response_for msg =
  let msg = String.strip msg in
  let is_question = String.is_suffix ~suffix:"?" msg in
  let is_silence = String.is_empty msg in
  let is_yelling =
    String.exists ~f:Char.is_uppercase msg &&
    not (String.exists ~f:Char.is_lowercase msg)
  in
  match is_question, is_yelling, is_silence with
  | true, true, _ -> "Calm down, I know what I'm doing!"
  | true, _, _-> "Sure."
  | _, true, _-> "Whoa, chill out!"
  | _, _, true -> "Fine. Be that way!"
  | _ -> "Whatever."

File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import "unicode"
import "strings"

func isYelling(remark string) bool {
	hasLetters := (strings.IndexFunc(remark, unicode.IsLetter) >= 0)
	isUpper := (strings.ToUpper(remark) == remark)
	return hasLetters && isUpper
}

func isQuestion(remark string) bool { return strings.HasSuffix(remark, "?") }

func isSilence(remark string) bool { return remark == "" }

func Hey(remark string) string {
	sanitized := strings.TrimSpace(remark)

	question := isQuestion(sanitized)
	yell := isYelling(sanitized)
	silence := isSilence(sanitized)

	switch {
		case silence : return "Fine. Be that way!"
		case yell && question: return "Calm down, I know what I'm doing!"
		case yell: return "Whoa, chill out!"
		case question: return "Sure."
		default: return "Whatever."
	}
}
